components/eventing-controller/pkg/backend/jetstreamv2/jetstream_internal_unit_test.go

```diff
@@ -0,0 +1,425 @@
+					jsCtx.AssertCalled(t, "Subscribe", jsSubject, mock.AnythingOfType("nats.MsgHandler"), mock.AnythingOfType("nats.subOptFn"))
+
+				}
+			},
+		},
+		{
+			name:              "Bind invalid NATS Subscription should result into error",
+			givenSubscription: subWithOneType,
+			givenMocks: func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier) {
+				for _, eventType := range sub.Status.Types {
+					jsSubject, jsSubKey, newConsumer, natsSub := generateConsumerInfra(jsBackend, sub, eventType)
+
+					// inject invalid subscriber
+					sub := &mocks.Subscriber{}
+					sub.On("IsValid").Return(false)
+					jsBackend.subscriptions[jsSubKey] = sub
+
+					// mock the expected calls
+					jsCtx.On("ConsumerInfo", jsBackend.Config.JSStreamName, jsSubKey.ConsumerName()).Return(newConsumer, nil)
+					jsCtx.On("Subscribe", jsSubject, mock.AnythingOfType("nats.MsgHandler"), mock.AnythingOfType("nats.subOptFn")).Return(natsSub, nats.ErrJetStreamNotEnabled)
+				}
+			},
+			givenAssertions: func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier) {
+				// check if the expected methods got called
+				for _, eventType := range sub.Status.Types {
+					jsSubject := jsBackend.getJetStreamSubject(sub.Spec.Source, eventType.CleanType, sub.Spec.TypeMatching)
+					jsSubKey := NewSubscriptionSubjectIdentifier(sub, jsSubject)
+
+					jsCtx.AssertCalled(t, "ConsumerInfo", jsBackend.Config.JSStreamName, jsSubKey.ConsumerName())
+					jsCtx.AssertNotCalled(t, "AddConsumer", jsBackend.Config.JSStreamName, mock.AnythingOfType("*nats.ConsumerConfig"))
+					jsCtx.AssertCalled(t, "Subscribe", jsSubject, mock.AnythingOfType("nats.MsgHandler"), mock.AnythingOfType("nats.subOptFn"))
+
+				}
+			},
+			wantErr: &backenderrors.FailedToSubscribeOnNATSError{OriginalError: nats.ErrJetStreamNotEnabled},
+		},
+		{
+			name:              "Update maxInFlight on the consumer side should be successful, when the subscription config changes",
+			givenSubscription: subWithOneType,
+			givenSubscriptionMap: map[SubscriptionSubjectIdentifier]Subscriber{
+				jsSubKey: subscriberMock,
+			},
+			givenMocks: func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier) {
+				for _, eventType := range sub.Status.Types {
+					jsSubject := jsBackend.getJetStreamSubject(sub.Spec.Source, eventType.CleanType, sub.Spec.TypeMatching)
+					jsSubKey := NewSubscriptionSubjectIdentifier(sub, jsSubject)
+					subscriberMock.On("IsValid").Return(true)
+
+					subWithOneType.Spec.Config[v1alpha2.MaxInFlightMessages] = "20"
+					jsCtx.On("ConsumerInfo", jsBackend.Config.JSStreamName, jsSubKey.ConsumerName()).Return(defaultConsumerInfo, nil)
+					jsCtx.On("UpdateConsumer", jsBackend.Config.JSStreamName, mock.AnythingOfType("*nats.ConsumerConfig")).Return(defaultConsumerInfo, nil)
+				}
+			},
```

Comment:
I see that the for loop creates quite some complexity for the test. It is in `givenMocks`, `givenAssertions` and the table-test body.

I have 2 ideas in my mind to work around this:
- 1. We have the subscription types under our control in the test. So why do we even need to pass multiple types ? 
- 2. What if we refactor `syncNATSConsumers` in such a way that there is only the for-loop and then an inner function call (say `processNATSSubject` e.g.,). We could then test both pieces of code in isolation. 
  - case 1: ensure that for each event type the the function (processNATSSubject) is called
  - case 2: ensure the correctness of the inner function (by only providing a single subject)

Reply:
I refactored the tests